docs/proposals/message-proto-design.md

```diff
@@ -0,0 +1,296 @@
+   * [Solve the filed index problem of Resource](#solve-the-filed-index-problem-of-resource)
+   * [Consider compatibility](#consider-compatibility)
+* [How to use in KubeEdge project](#how-to-use-in-KubeEdge-project)
+* [Which projects need to be modified](#which-projects-need-to-be-modified)
+
+## Abstract 
+
+currently, In beehive framework, model message is defined and serialized using json format. the performance is slower than protobuf.
+
+and for parsing the message resource, each field index is hardcoded. such as (https://github.com/kubeedge/beehive/blob/master/pkg/common/util/parse_resource.go)
+
+in KubeEdge project , we can see many `Split(string, "/")` function which to parse the resource path string.  these are hard code.   when we deal will the message resource after we receive message  . we need to know the format of the assembly In advance。 then use `Split()` function to get the field we want.    it`s not easy to expand. It's not so extensible. so we need a more extended form.
+
+please see the #399 issue of kubeeedge 
+
+in addition, currently  when we use quic or websocket protocol to paas message, we need convert model message to probuf message (which defined in `github.com/kubeedge/viaduct/pkg/protos/message/message.pb.go`). and then send .  This also affects the efficiency of the transmission, rather than simply turning the model mesage into proto message。
+
+## Need to do
+
+*  Change the model message to protobuf message  directly
+*  Solve the filed index problem of `Resource`
+*  Consider compatibility  
+
+## How to do 
+
+##  Change the model message to protobuf message  directly
+
+create proto file in `model` dir to define message struct. and use `go generate`(will move to makefile) to create message.pb.go file 
+
+```
+syntax = "proto3";
+package model;
+import "google/protobuf/any.proto";
+
+message MessageHeader {
+    string ID = 1;
+    string ParentID = 2;
+    int64 Timestamp = 3;
+    bool Sync = 4;
+}
+message MessageRoute {
+    string Source = 1;
+    string Group = 2;
+    string Operation = 3;
+    google.protobuf.Any Resource = 4;
+}
+
+message Message {
+    MessageHeader Header = 1;
+    MessageRoute Router = 2;
+    google.protobuf.Any Data = 3;
+}
+```
+
+we use `Any type` to define `Resource` and `Data` field.  `Data` field is the original `Content` field.
+
+##  Solve the filed index problem of `Resource`
+
+this problem is the most important, currently, beehive define path string to stand for resource, such as `/node/{nodeID}/{namespace}/{resourceType}/{resource}`, different path stand for different resource.  but we need hardcord to parse `nodeid` ,`resourceType`,`resourceid`. It's not so extensible, so we need a more extended form. we need change `Resource` to golang `interface{}`. but in protobuf, there is not `interface` define. we need use `Any` type.
+
+we can see `MessageRoute ` definition.
+
+we use `github.com/gogo/protobuf`  package not offical `github.com/golang/protobuf` package . 
```

Comment:
Pelease add reasons(benifits) of why gogo/protobuf is suggested here instead of golang/protobuf